File: implementations/go/r57.go
package b57

import (
	"crypto/rand"
	"encoding/binary"
	"errors"
	"sync"
	"sync/atomic"
	"time"

	"lukechampine.com/blake3"
)

var randRead = rand.Read

var r57Counter atomic.Uint32

var (
	r57SessionSecretOnce sync.Once
	r57SessionSecret     [32]byte
	r57SessionSecretErr  error

	r57DeviceSecretOnce sync.Once
	r57DeviceSecret     [32]byte
	r57DeviceSecretErr  error

	r57MasterSecretOnce sync.Once
	r57MasterSecret     [32]byte
	r57MasterSecretErr  error
)

// R57Mode defines the random generation mode for R57 identifiers.
type R57Mode int

const (
	// R57ModeCSPRNG is the default mode, generating 128-bit entropy via CSPRNG.
	R57ModeCSPRNG R57Mode = 1
	// R57ModeHashEntropy generates deterministic randomness from high-entropy source.
	R57ModeHashEntropy R57Mode = 2
	// R57ModeKDFDerived generates deterministic randomness via KDF.
	R57ModeKDFDerived R57Mode = 3
	// R57ModeCounterKDF generates sequence randomness via KDF and counter.
	R57ModeCounterKDF R57Mode = 4
	// R57ModeTimestampKDF generates time-bound randomness via KDF.
	R57ModeTimestampKDF R57Mode = 5
	// R57ModeHardwareRNG generates randomness from a hardware source.
	R57ModeHardwareRNG R57Mode = 6
	// R57ModeUUIDv4Compat provides UUIDv4 compatibility.
	R57ModeUUIDv4Compat R57Mode = 7
	// R57ModeHybridEntropy generates randomness mixing multiple sources.
	R57ModeHybridEntropy R57Mode = 8
)

// R57Generate creates a new R57 identifier using the specified mode.
// It generates exactly 128-bit entropy and returns a 22-character B57 string.
func R57Generate(mode R57Mode) (string, error) {
	raw, err := generateR57Entropy(mode)
	if err != nil {
		return "", err
	}
	return encodeR57128(raw), nil
}

func generateR57Entropy(mode R57Mode) ([]byte, error) {
	switch mode {
	case R57ModeCSPRNG:
		return readEntropyBytes(16)
	case R57ModeHashEntropy:
		return generateHashEntropy()
	case R57ModeKDFDerived:
		return generateKDFDerivedEntropy()
	case R57ModeCounterKDF:
		return generateCounterKDFEntropy()
	case R57ModeTimestampKDF:
		return generateTimestampKDFEntropy()
	case R57ModeHardwareRNG:
		// No hardware-specific source is exposed here; fall back to the system CSPRNG.
		return readEntropyBytes(16)
	case R57ModeUUIDv4Compat:
		return generateUUIDv4CompatEntropy()
	case R57ModeHybridEntropy:
		return generateHybridEntropy()
	default:
		return nil, NewInvalidModeError(int(mode))
	}
}

func generateHashEntropy() ([]byte, error) {
	seed, err := readEntropyBytes(16)
	if err != nil {
		return nil, err
	}
	return mixEntropy128(seed), nil
}

func generateKDFDerivedEntropy() ([]byte, error) {
	context, err := readEntropyBytes(16)
	if err != nil {
		return nil, err
	}
	secret, err := r57GetSecret(&r57MasterSecretOnce, &r57MasterSecret, &r57MasterSecretErr)
	if err != nil {
		return nil, err
	}
	return mixEntropy128(secret, context), nil
}

func generateCounterKDFEntropy() ([]byte, error) {
	randomComponent, err := readEntropyBytes(12)
	if err != nil {
		return nil, err
	}
	counter := make([]byte, 4)
	binary.BigEndian.PutUint32(counter, r57Counter.Add(1))
	return mixEntropy128(counter, randomComponent), nil
}

func generateTimestampKDFEntropy() ([]byte, error) {
	randomComponent, err := readEntropyBytes(12)
	if err != nil {
		return nil, err
	}
	timestamp := make([]byte, 4)
	binary.BigEndian.PutUint32(timestamp, uint32(time.Now().Unix()/60))
	secret, err := r57GetSecret(&r57DeviceSecretOnce, &r57DeviceSecret, &r57DeviceSecretErr)
	if err != nil {
		return nil, err
	}
	return mixEntropy128(secret, timestamp, randomComponent), nil
}

func generateUUIDv4CompatEntropy() ([]byte, error) {
	raw, err := readEntropyBytes(16)
	if err != nil {
		return nil, err
	}
	raw[6] = (raw[6] & 0x0F) | 0x40
	raw[8] = (raw[8] & 0x3F) | 0x80
	return raw, nil
}

func generateHybridEntropy() ([]byte, error) {
	seed, err := readEntropyBytes(16)
	if err != nil {
		return nil, err
	}
	counter := make([]byte, 4)
	binary.BigEndian.PutUint32(counter, r57Counter.Add(1))
	timestamp := make([]byte, 8)
	binary.BigEndian.PutUint64(timestamp, uint64(time.Now().UnixNano()))
	secret, err := r57GetSecret(&r57SessionSecretOnce, &r57SessionSecret, &r57SessionSecretErr)
	if err != nil {
		return nil, err
	}
	return mixEntropy128(seed, counter, timestamp, secret), nil
}

func readEntropyBytes(length int) ([]byte, error) {
	buf := make([]byte, length)
	n, err := randRead(buf)
	if err != nil {
		return nil, err
	}
	if n != length {
		return nil, errors.New("failed to read 16 bytes of entropy")
	}
	return buf, nil
}

func r57GetSecret(once *sync.Once, target *[32]byte, targetErr *error) ([]byte, error) {
	once.Do(func() {
		buf, err := readEntropyBytes(len(target))
		if err != nil {
			*targetErr = err
			return
		}
		copy(target[:], buf)
	})
	if *targetErr != nil {
		return nil, *targetErr
	}
	return target[:], nil
}

func mixEntropy128(parts ...[]byte) []byte {
	hasher := blake3.New(32, nil)
	for _, part := range parts {
		_, _ = hasher.Write(part)
	}
	sum := hasher.Sum(nil)
	return append([]byte(nil), sum[:16]...)
}

func encodeR57128(raw []byte) string {
	encoded := Encode(raw)
	if len(encoded) == 22 {
		return encoded
	}

	derived := append([]byte(nil), raw...)
	for len(encoded) < 22 {
		derived = mixEntropy128(derived, []byte{0x57})
		encoded = Encode(derived)
	}
	return encoded
}

// R57IsValid checks if the provided string is a valid R57 identifier
// (length 22 and contains only valid B57 characters).
func R57IsValid(s string) bool {
	if len(s) != 22 {
		return false
	}
	return IsValid(s)
}

// R57IsCanonical verifies if the output string is canonical according to R57 and B57.
func R57IsCanonical(s string) bool {
	if len(s) != 22 {
		return false
	}
	return IsCanonical(s)
}

File: implementations/ts/r57.ts
import { randomBytes } from 'node:crypto';
import { blake3 } from '@noble/hashes/blake3';
import { encode, isValid, isCanonical } from './b57.js';
import { newInvalidR57ModeError } from './errors.js';

let r57Counter = 0;
let sessionSecret;
let deviceSecret;
let masterSecret;

export const R57Mode = Object.freeze({
  R57_1_CSPRNG: 1,
  R57_2_HASH_ENTROPY: 2,
  R57_3_KDF_DERIVED: 3,
  R57_4_COUNTER_KDF: 4,
  R57_5_TIMESTAMP_KDF: 5,
  R57_6_HARDWARE_RNG: 6,
  R57_7_UUIDV4_COMPAT: 7,
  R57_8_HYBRID_ENTROPY: 8
});

export function r57Generate(mode) {
  const raw = generateR57Entropy(mode);
  return encodeR57128(raw);
}

function generateR57Entropy(mode) {
  switch (mode) {
    case R57Mode.R57_1_CSPRNG:
      return readEntropyBytes(16);
    case R57Mode.R57_2_HASH_ENTROPY:
      return mixEntropy128(readEntropyBytes(16));
    case R57Mode.R57_3_KDF_DERIVED:
      return generateKDFDerivedEntropy();
    case R57Mode.R57_4_COUNTER_KDF:
      return generateCounterKDFEntropy();
    case R57Mode.R57_5_TIMESTAMP_KDF:
      return generateTimestampKDFEntropy();
    case R57Mode.R57_6_HARDWARE_RNG:
      return readEntropyBytes(16);
    case R57Mode.R57_7_UUIDV4_COMPAT:
      return generateUUIDv4CompatEntropy();
    case R57Mode.R57_8_HYBRID_ENTROPY:
      return generateHybridEntropy();
    default:
      throw newInvalidR57ModeError(mode);
  }
}

function generateKDFDerivedEntropy() {
  if (!masterSecret) {
    masterSecret = readEntropyBytes(32);
  }
  const context = readEntropyBytes(16);
  return mixEntropy128(masterSecret, context);
}

function generateCounterKDFEntropy() {
  const randomComponent = readEntropyBytes(12);
  r57Counter += 1;
  const counter = new Uint8Array(4);
  const view = new DataView(counter.buffer);
  view.setUint32(0, r57Counter >>> 0, false);
  return mixEntropy128(counter, randomComponent);
}

function generateTimestampKDFEntropy() {
  if (!deviceSecret) {
    deviceSecret = readEntropyBytes(32);
  }
  const randomComponent = readEntropyBytes(12);
  const timestamp = new Uint8Array(4);
  const view = new DataView(timestamp.buffer);
  view.setUint32(0, Math.floor(Date.now() / 60000) >>> 0, false);
  return mixEntropy128(deviceSecret, timestamp, randomComponent);
}

function generateUUIDv4CompatEntropy() {
  const raw = readEntropyBytes(16);
  raw[6] = (raw[6] & 0x0f) | 0x40;
  raw[8] = (raw[8] & 0x3f) | 0x80;
  return raw;
}

function generateHybridEntropy() {
  if (!sessionSecret) {
    sessionSecret = readEntropyBytes(32);
  }
  const seed = readEntropyBytes(16);
  r57Counter += 1;

  const counter = new Uint8Array(4);
  new DataView(counter.buffer).setUint32(0, r57Counter >>> 0, false);

  const timestamp = new Uint8Array(8);
  const now = BigInt(Date.now()) * 1000000n;
  new DataView(timestamp.buffer).setBigUint64(0, now, false);

  return mixEntropy128(seed, counter, timestamp, sessionSecret);
}

function readEntropyBytes(length) {
  return new Uint8Array(randomBytes(length));
}

function mixEntropy128(...parts) {
  const totalLength = parts.reduce((sum, p) => sum + p.length, 0);
  const combined = new Uint8Array(totalLength);
  let offset = 0;
  for (const part of parts) {
    combined.set(part, offset);
    offset += part.length;
  }
  return blake3(combined, { dkLen: 16 });
}

function encodeR57128(raw) {
  let encoded = encode(raw);
  if (encoded.length === 22) {
    return encoded;
  }

  let derived = raw;
  while (encoded.length < 22) {
    derived = mixEntropy128(derived, new Uint8Array([0x57]));
    encoded = encode(derived);
  }
  return encoded;
}

export function r57IsValid(s) {
  if (typeof s !== 'string') {
    return false;
  }
  if (s.length !== 22) {
    return false;
  }
  return isValid(s);
}

export function r57IsCanonical(s) {
  if (typeof s !== 'string') {
    return false;
  }
  if (s.length !== 22) {
    return false;
  }
  return isCanonical(s);
}
